File: bindings/go/pkg/transport/http/custom_endpoint_test.go
package http

import (
	"testing"
	"time"
)

// Test custom endpoint configuration
func TestClientConfig_CustomPath(t *testing.T) {
	config := ClientConfig{
		BaseURL: "http://localhost:8000",
		Path:    "/umicp", // Custom path for Vectorizer
		Timeout: 30 * time.Second,
	}

	if config.Path != "/umicp" {
		t.Errorf("Expected path /umicp, got %s", config.Path)
	}
}

// Test default path
func TestClientConfig_DefaultPath(t *testing.T) {
	config := DefaultClientConfig()

	// Default path should be /umicp
	if config.Path != "/umicp" {
		t.Errorf("Expected default path /umicp, got %s", config.Path)
	}
}

// Test Vectorizer endpoint configuration
func TestClientConfig_VectorizerEndpoint(t *testing.T) {
	config := ClientConfig{
		BaseURL: "http://localhost:8000",
		Path:    "/umicp",
		Timeout: 30 * time.Second,
	}

	client := NewClient(config)
	if client == nil {
		t.Fatal("Failed to create client")
	}

	// Verify path is set
	if client.config.Path != "/umicp" {
		t.Errorf("Expected path /umicp, got %s", client.config.Path)
	}
}

// Test standard endpoint configuration
func TestClientConfig_StandardEndpoint(t *testing.T) {
	config := ClientConfig{
		BaseURL: "http://localhost:9000",
		Path:    "/message",
		Timeout: 30 * time.Second,
	}

	client := NewClient(config)
	if client == nil {
		t.Fatal("Failed to create client")
	}

	// Verify path is set
	if client.config.Path != "/message" {
		t.Errorf("Expected path /message, got %s", client.config.Path)
	}
}

// Test empty path defaults to /umicp
func TestClientConfig_EmptyPathUsesDefault(t *testing.T) {
	config := ClientConfig{
		BaseURL: "http://localhost:8000",
		Path:    "", // Empty path
	}

	client := NewClient(config)
	if client == nil {
		t.Fatal("Failed to create client")
	}

	// Should use default /umicp
	if client.config.Path != "/umicp" {
		t.Errorf("Expected default path /umicp, got %s", client.config.Path)
	}
}

// Test multiple clients with different endpoints
func TestClient_MultipleEndpoints(t *testing.T) {
	// Client for Vectorizer
	vectorizerConfig := ClientConfig{
		BaseURL: "http://localhost:8000",
		Path:    "/umicp",
	}
	vectorizerClient := NewClient(vectorizerConfig)

	// Client for standard server
	standardConfig := ClientConfig{
		BaseURL: "http://localhost:9000",
		Path:    "/message",
	}
	standardClient := NewClient(standardConfig)

	// Both should be independent
	if vectorizerClient.config.Path == standardClient.config.Path {
		t.Error("Clients should have different paths")
	}

	if vectorizerClient.config.BaseURL == standardClient.config.BaseURL {
		t.Error("Clients should have different base URLs")
	}
}

// Test client creation with custom path
func TestNewClient_WithCustomPath(t *testing.T) {
	config := ClientConfig{
		BaseURL: "http://localhost:8000",
		Path:    "/custom/endpoint",
	}

	client := NewClient(config)
	if client == nil {
		t.Fatal("Failed to create client")
	}

	if client.config.Path != "/custom/endpoint" {
		t.Errorf("Expected path /custom/endpoint, got %s", client.config.Path)
	}
}

// Test backward compatibility - old code still works
func TestClient_BackwardCompatibility(t *testing.T) {
	// Old code without Path field
	config := ClientConfig{
		BaseURL: "http://localhost:8000",
		// Path not specified - should use default
	}

	client := NewClient(config)
	if client == nil {
		t.Fatal("Failed to create client")
	}

	// Should use default
	if client.config.Path != "/umicp" {
		t.Errorf("Expected default path /umicp for backward compatibility, got %s", client.config.Path)
	}
}

// Test version 0.2.2 supports custom endpoints
func TestVersion_0_2_2_SupportsCustomEndpoints(t *testing.T) {
	// Test that clients can be created with different endpoints
	config1 := ClientConfig{
		BaseURL: "http://localhost:8000",
		Path:    "/umicp",
	}
	client1 := NewClient(config1)

	config2 := ClientConfig{
		BaseURL: "http://localhost:9000",
		Path:    "/message",
	}
	client2 := NewClient(config2)

	if client1 == nil || client2 == nil {
		t.Fatal("Failed to create clients")
	}

	// Verify paths are different
	if client1.config.Path == client2.config.Path {
		t.Error("Clients should support different paths")
	}
}

// Test config validation initializes empty path
func TestClient_ConfigValidation(t *testing.T) {
	config := ClientConfig{
		BaseURL: "http://localhost:8000",
		// All other fields zero/empty
	}

	client := NewClient(config)
	if client == nil {
		t.Fatal("Failed to create client")
	}

	// Path should be initialized
	if client.config.Path == "" {
		t.Error("Path should be initialized to default")
	}

	// Timeout should be initialized
	if client.config.Timeout == 0 {
		t.Error("Timeout should be initialized to default")
	}
}

File: bindings/typescript/test/custom-endpoint.test.ts
/**
 * Custom Endpoint Support Tests (v0.2.2)
 *
 * Verifies that the TypeScript UMICP implementation supports
 * custom endpoint paths for compatibility with different servers
 * (e.g., Vectorizer uses /umicp, standard servers use /message)
 */

import { describe, it, expect } from 'vitest';
import { StreamableHTTPClient, type StreamableHTTPClientOptions } from '../src/transports/streamable-http.js';

describe('Custom Endpoint Support', () => {
  describe('Client Configuration', () => {
    it('should support default path /umicp', () => {
      const options: StreamableHTTPClientOptions = {
        baseUrl: 'http://localhost:8000',
      };

      const client = new StreamableHTTPClient(options);
      expect(client).toBeDefined();
      // Default path should be /umicp
    });

    it('should support custom path for Vectorizer', () => {
      const options: StreamableHTTPClientOptions = {
        baseUrl: 'http://localhost:8000',
        path: '/umicp',
      };

      const client = new StreamableHTTPClient(options);
      expect(client).toBeDefined();
    });

    it('should support custom path for standard server', () => {
      const options: StreamableHTTPClientOptions = {
        baseUrl: 'http://localhost:9000',
        path: '/message',
      };

      const client = new StreamableHTTPClient(options);
      expect(client).toBeDefined();
    });

    it('should support empty path', () => {
      const options: StreamableHTTPClientOptions = {
        baseUrl: 'http://localhost:8000',
        path: '',
      };

      const client = new StreamableHTTPClient(options);
      expect(client).toBeDefined();
    });

    it('should support complex path', () => {
      const options: StreamableHTTPClientOptions = {
        baseUrl: 'http://localhost:8000',
        path: '/api/v1/umicp',
      };

      const client = new StreamableHTTPClient(options);
      expect(client).toBeDefined();
    });
  });

  describe('Multiple Clients', () => {
    it('should support multiple clients with different endpoints', () => {
      const vectorizerClient = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000',
        path: '/umicp',
      });

      const standardClient = new StreamableHTTPClient({
        baseUrl: 'http://localhost:9000',
        path: '/message',
      });

      expect(vectorizerClient).toBeDefined();
      expect(standardClient).toBeDefined();
      expect(vectorizerClient).not.toBe(standardClient);
    });

    it('should allow same base URL with different paths', () => {
      const client1 = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000',
        path: '/umicp',
      });

      const client2 = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000',
        path: '/message',
      });

      expect(client1).toBeDefined();
      expect(client2).toBeDefined();
    });
  });

  describe('Backward Compatibility', () => {
    it('should work without specifying path (uses default)', () => {
      // Old code that doesn't specify path
      const client = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000',
      });

      expect(client).toBeDefined();
      expect(client.isConnected()).toBe(false);
    });

    it('should maintain default behavior', () => {
      const client1 = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000',
      });

      const client2 = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000',
        path: '/umicp', // Explicit default
      });

      expect(client1).toBeDefined();
      expect(client2).toBeDefined();
    });
  });

  describe('Edge Cases', () => {
    it('should handle path with trailing slash', () => {
      const client = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000',
        path: '/umicp/',
      });

      expect(client).toBeDefined();
    });

    it('should handle base URL with trailing slash', () => {
      const client = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000/',
        path: '/umicp',
      });

      expect(client).toBeDefined();
    });

    it('should handle both with trailing slashes', () => {
      const client = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000/',
        path: '/umicp/',
      });

      expect(client).toBeDefined();
    });

    it('should handle path with query parameters', () => {
      const client = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000',
        path: '/umicp?version=1',
      });

      expect(client).toBeDefined();
    });
  });

  describe('Configuration Options', () => {
    it('should support all options with custom path', () => {
      const client = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000',
        path: '/umicp',
        timeout: 30000,
        headers: {
          'X-Custom-Header': 'value',
        },
      });

      expect(client).toBeDefined();
    });

    it('should support HTTPS endpoints', () => {
      const client = new StreamableHTTPClient({
        baseUrl: 'https://localhost:8443',
        path: '/umicp',
      });

      expect(client).toBeDefined();
    });
  });

  describe('Version 0.2.2 Features', () => {
    it('should support custom endpoints in v0.2.2', () => {
      // Test that v0.2.2 supports custom endpoints
      const vectorizerClient = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000',
        path: '/umicp',
      });

      const messageClient = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000',
        path: '/message',
      });

      expect(vectorizerClient).toBeDefined();
      expect(messageClient).toBeDefined();
    });

    it('should maintain connection state interface', () => {
      const client = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000',
        path: '/umicp',
      });

      // Client should have isConnected method
      expect(typeof client.isConnected).toBe('function');
      expect(client.isConnected()).toBe(false);
    });

    it('should maintain disconnect interface', () => {
      const client = new StreamableHTTPClient({
        baseUrl: 'http://localhost:8000',
        path: '/umicp',
      });

      // Client should have disconnect method
      expect(typeof client.disconnect).toBe('function');

      // Should not throw
      expect(() => client.disconnect()).not.toThrow();
    });
  });

  describe('Type Safety', () => {
    it('should accept valid StreamableHTTPClientOptions', () => {
      const options: StreamableHTTPClientOptions = {
        baseUrl: 'http://localhost:8000',
        path: '/umicp',
        timeout: 30000,
        headers: {
          'Content-Type': 'application/json',
        },
      };

      const client = new StreamableHTTPClient(options);
      expect(client).toBeDefined();
    });

    it('should work with minimal options', () => {
      const options: StreamableHTTPClientOptions = {
        baseUrl: 'http://localhost:8000',
      };

      const client = new StreamableHTTPClient(options);
      expect(client).toBeDefined();
    });
  });
});

describe('StreamableHTTPClient Path Handling', () => {
  it('should use provided path in requests', () => {
    const client = new StreamableHTTPClient({
      baseUrl: 'http://localhost:8000',
      path: '/custom',
    });

    expect(client).toBeDefined();
    expect(client.isConnected()).toBe(false);
  });

  it('should support Vectorizer configuration', () => {
    // Typical Vectorizer configuration
    const client = new StreamableHTTPClient({
      baseUrl: 'http://localhost:8000',
      path: '/umicp',
      timeout: 30000,
    });

    expect(client).toBeDefined();
  });

  it('should support standard UMICP server configuration', () => {
    // Standard UMICP server configuration
    const client = new StreamableHTTPClient({
      baseUrl: 'http://localhost:9000',
      path: '/message',
      timeout: 30000,
    });

    expect(client).toBeDefined();
  });
});

